File: java/20. Valid Parentheses/Solution.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        Map<Character, Character> mapping = new HashMap<>(Map.of(')', '(', '}', '{', ']', '['));
        for (char c : s.toCharArray()) {
            if (mapping.containsKey(c)) {
                Character last = stack.isEmpty() ? '.' : stack.pop();
                if (!mapping.get(c).equals(last)) return false;
            } else {
                stack.add(c);
            }
        }

        return stack.isEmpty();
    }
}

File: python/20. Valid Parentheses/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        mapping = {')' : '(', '}' : '{', ']' : '['}
        for char in s:
            if char in mapping:
                last = stack.pop() if stack else '.'
                if last != mapping.get(char):
                    return False
                else:
                    stack.append(char)

        return not stack